crypto/http/http_client.c

```diff
@@ -0,0 +1,1200 @@
+ * After disconnect the modified BIO will be deallocated using BIO_free_all().
+ */
+BIO *OSSL_HTTP_transfer(const char *server, const char *port, const char *path,
+                        int use_ssl, const char *proxy, const char *proxy_port,
+                        OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
+                        const STACK_OF(CONF_VALUE) *headers,
+                        const char *content_type, BIO *req_mem,
+                        int maxline, unsigned long max_resp_len, long timeout,
+                        const char *expected_ct, int expect_asn1,
+                        char **redirection_url)
+{
+    time_t start_time = timeout > 0 ? time(NULL) : 0;
+    BIO *bio;
+    OSSL_HTTP_REQ_CTX *rctx;
+    BIO *resp = NULL;
+
+    if (redirection_url != NULL)
+        *redirection_url = NULL; /* do this beforehand to prevent dbl free */
+
+    if (use_ssl && bio_update_fn == NULL) {
+        HTTPerr(0, HTTP_R_TLS_NOT_ENABLED);
+        return NULL;
+    }
+    /* remaining parameters are checked indirectly by the functions called */
+
+    if ((bio = HTTP_new_bio(server, port, proxy, proxy_port)) == NULL)
+        return NULL;
+
+    (void)ERR_set_mark(); /* prepare removing any spurious libssl errors */
+    if (BIO_connect_retry(bio, timeout) <= 0)
+        goto end;
+    /* now timeout is guaranteed to be >= 0 */
+
+    /* callback can be used to wrap or prepend TLS session */
+    if (bio_update_fn != NULL) {
+        BIO *orig_bio = bio;
+        bio = (*bio_update_fn)(bio, arg, 1 /* connect */, use_ssl);
+        if (bio == NULL) {
+            bio = orig_bio;
+            goto end;
+        }
+    }
+
+    rctx = HTTP_REQ_CTX_new(bio, !use_ssl && proxy != NULL, server, port, path,
+                            headers, content_type, req_mem, maxline,
+                            max_resp_len, update_timeout(timeout, start_time),
+                            expected_ct, expect_asn1);
+    if (rctx == NULL)
+        goto end;
+
+    resp = OSSL_HTTP_REQ_CTX_transfer(rctx);
+    if (resp == NULL) {
+        if (rctx->redirection_url != NULL) {
+            if (redirection_url == NULL)
+                HTTPerr(0, HTTP_R_REDIRECTION_NOT_ENABLED);
+            else
+                /* may be NULL if out of memory: */
+                (*redirection_url) = OPENSSL_strdup(rctx->redirection_url);
+        } else {
+            char buf[200];
+            unsigned long err = ERR_peek_error();
+            int lib = ERR_GET_LIB(err);
+            int reason = ERR_GET_REASON(err);
+            if (lib == ERR_LIB_SSL || lib == ERR_LIB_HTTP
+                    || (lib == ERR_LIB_BIO && reason == BIO_R_CONNECT_TIMEOUT)
+                    || (lib == ERR_LIB_BIO && reason == BIO_R_CONNECT_ERROR)
+                    || (lib == ERR_LIB_CMP
+                        && reason == CMP_R_POTENTIALLY_INVALID_CERTIFICATE)) {
+                BIO_snprintf(buf, 200, "server=%s:%s", server, port);
```

Comment:
how buf is terminated with an '\0' in any cases?

Reply:
This should be done by `BIO_snprintf`.

Reply:
As discussed in person, the OpenSSL documentation appears to imply this but unfortunately it is rather vague.
I've just tried verifying this in the implementation of `BIO_snprintf`, but this is pretty complex and indirect. So I've resorted to checking current "best practice" of using this function in various other places in `crypto/*/`, and there it is used in the same way as done here.

Reply:
OK